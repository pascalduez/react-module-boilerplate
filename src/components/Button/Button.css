.root {
  padding: 1rem 2rem;
  border: 1px solid;
  background-color: transparent;
  cursor: pointer;
}

.primary {
  color: var(--color-primary);
}

.secondary {
  color: var(--color-secondary);
}

.danger {
  color: var(--color-danger);
}
