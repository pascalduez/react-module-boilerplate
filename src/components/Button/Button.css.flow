// @flow

declare module.exports: $ReadOnly<{
  root: string,
  primary: string,
  secondary: string,
  danger: string,
}>;
