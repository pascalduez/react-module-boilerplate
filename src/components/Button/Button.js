/* @flow */
/* eslint css-modules/no-unused-class: [2, { markAsUsed: ['primary', 'secondary', 'danger'] }] */
/* eslint react/default-props-match-prop-types: ['error', { 'allowRequiredDefaults': true }] */
/* eslint-disable react/button-has-type */

import * as React from 'react';
import clsx from 'clsx';

import styles from './Button.css';

export type Props = {|
  children: React.Node,
  theme?: 'primary' | 'secondary' | 'danger',
  type?: 'submit' | 'button' | 'reset',
  className?: string,
  onClick?: (evt: SyntheticEvent<HTMLButtonElement>) => mixed,
|};

function Button({
  theme = 'primary',
  type = 'submit',
  className,
  ...extraProps
}: Props) {
  return (
    <button
      type={type}
      className={clsx(styles.root, styles[theme], className)}
      {...extraProps}
    />
  );
}

export default Button;
