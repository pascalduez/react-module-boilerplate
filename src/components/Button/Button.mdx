---
name: Button
---

import { Playground, PropsTable } from 'docz';
import Button from './Button';

# Button

## Basic usage

<Playground>
  <Button>Click me</Button>
  <Button theme="secondary">Click me</Button>
  <Button theme="danger">Click me</Button>
</Playground>

## PropTypes

<PropsTable of={Button} />
