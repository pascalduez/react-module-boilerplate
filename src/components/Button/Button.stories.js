// @flow

import * as React from 'react';

import Button from './Button';

export default {
  component: Button,
  title: 'Button',
};

export const primary = () => <Button>Test</Button>;

export const secondary = () => <Button theme="secondary">Test</Button>;

export const danger = () => <Button theme="danger">Test</Button>;
