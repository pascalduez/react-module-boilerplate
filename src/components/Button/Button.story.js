// @flow

import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, text, select } from '@storybook/addon-knobs/react';
import { withInfo } from '@storybook/addon-info';
import { checkA11y } from '@storybook/addon-a11y';

import Button from './Button';

const themeOptions = {
  primary: 'primary',
  secondary: 'secondary',
  danger: 'danger',
};

storiesOf('Button', module)
  .addDecorator(withKnobs)
  .addDecorator(checkA11y)
  .add('label', () => <Button>{text('Label property', 'Hello')}</Button>)
  .add('theme property', () => (
    <Button theme={select('Theme property', themeOptions, 'primary')}>
      Hello
    </Button>
  ))
  .add('handleClick property', () => (
    <Button handleClick={action('Hello from click handler')}>Click me!</Button>
  ))
  .add(
    'Default usage',
    withInfo({
      text:
        'This is the default Button usage, just providing a label as children.',
      inline: true,
      propTables: [Button],
    })(() => <Button>Hello</Button>)
  );
