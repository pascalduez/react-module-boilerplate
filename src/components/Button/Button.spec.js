/* @flow */
/* eslint css-modules/no-unused-class: [2, { markAsUsed: ['secondary', 'danger'] }] */

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event'

import Button from './Button';
import styles from './Button.css';

describe('<Button>', () => {
  it('should render the default markup', () => {
    render(<Button>Test</Button>);

    const button = screen.getByText('Test');

    expect(button.tagName).toBe('BUTTON');
  });

  it('should render the default classes', () => {
    render(<Button>Test</Button>);

    const button = screen.getByText('Test');

    expect(button).toHaveClass(styles.root);
    expect(button).toHaveClass(styles.primary);
  });

  // Which is equivalent to:

  it('should render the default markup and classes [snapshot]', () => {
    const { asFragment } = render(<Button>Test</Button>);

    expect(asFragment()).toMatchSnapshot();
  });

  // Event handlers.

  it('should trigger click event handlers', () => {
    const spy = jest.fn();
    render(<Button onClick={spy}>Test</Button>);

    const button = screen.getByText('Test');

    userEvent.click(button);

    expect(spy).toHaveBeenCalled();
  });
});
