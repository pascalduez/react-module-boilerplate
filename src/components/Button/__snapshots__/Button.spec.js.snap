// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> should render the default markup and classes [snapshot] 1`] = `
<DocumentFragment>
  <button
    class="Button-root Button-primary"
    type="submit"
  >
    Test
  </button>
</DocumentFragment>
`;
